### PR TITLE
 Fix Blendshapes so they are cleared from MeshInstance3D when mesh is changed. 

### LINKED ISSUE #96873: Blendshapes aren't cleared from MeshInstance3D when mesh is changed.
### Tested versions

Reproducible:
- v4.4.dev2.official [97ef3c837]
- v4.3.stable.official [77dcf97d8]
...Likely earlier too.

### System information

Godot v4.3.stable - Windows 10.0.22631 - GLES3 (Compatibility) - NVIDIA GeForce RTX 4090 (NVIDIA; 31.0.15.3699) - AMD Ryzen 9 7950X 16-Core Processor (32 Threads)

### Issue description

When changing MeshInstance3D mesh without clearing it first, the blendshapes of the previous mesh remain.
Before:

before

After changing:

after

Thus some errors are spammed:

> scene/3d/mesh_instance_3d.cpp:158 - Index p_blend_shape = 0 is out of bounds ((int)blend_shape_tracks.size() = 0).

### Steps to reproduce

Create MeshInstance3D, attach a mesh with blendshapes to it, change the mesh to an another (without clearing the mesh property).

**Edit**: As I'm writing this I realized I Forgot to test what happens when mesh is changed to an another with blendshapes will report shortly.

### Minimal reproduction project (MRP)

[meshswapissue.zip](https://github.com/user-attachments/files/16968957/meshswapissue.zip)

### PR DESCRIPTION
**Pull request by deppong**
Address the issue #96873 where blendshapes aren't properly removed when you swap out the mesh in a MeshInstance3D node. Seems like the blend_shape_tracks and blend_shape_properties needed to be cleared even if the mesh that changed was valid. 

This is my first PR for godot, so let me know if there is anything else I can do on my end. 

After making the changes, I then tested with the Minimal reproduction project (MRP) provided in the issue, and when the mesh is changed, the Blendshapes property is properly updated (and removed if no blendshapes are present in the new mesh)

current master commit (2c136e6170a40f58f2dfb89d32eadfca7156ef37) still has this issue when tested.

*Bugsquad edit:*
- Fixes #96873

### CODE REVIEW

**Comment by Lielay9** [comment]
Just woke up and did basic testing, too lazy to format so here's copy from txt (gltf2 used for importing; importing = importing different mesh with same filepath):

Before patch (4.3):
- Import Inherited scene -> Values preserved by name (Good)
- Import Save to file mesh -> Values not preserved (Not good)
- Import 2 to 1 shapes -> Values preserved by name (Good) but error:  
> scene/3d/mesh_instance_3d.cpp:158 - Index p_blend_shape = 1 is out of bounds ((int)blend_shape_tracks.size() = 1)

- Changing mesh from code (mesh_instance.mesh = new_mesh) preserves blendshape values. (Not good)
- Changing mesh from editor doesn't clear blendshapes (blendshapes from previous meshes compound). (Not good)

After Patch:
- Import Inherited scene -> Values preserved by name (Good)
- Import Save to file mesh -> Values not preserved (Not good; same as before)
- Import 2 to 1 shapes -> Values preserved by name (Good)
- Changing mesh from code (mesh_instance.mesh = new_mesh) clears blendshape values. (Good)
- Changing mesh from editor clears blendshapes (even if blendshapes with the same name exists) (Good?)

Here's the project used, just replace 'suzanne.glb' with any of the numbered ones from 'suzannes' directory to reimport.
[meshswapissue.zip](https://github.com/user-attachments/files/16981809/meshswapissue.zip)

I think the result is expected as the blendshape values are saved to the scene and not the resource. I haven't woken up yet properly so someone else can take a look too, but it seems like the importing works the same if not a bit better. Other improvements are out of scope of this pr.

**Comment by deppong** [comment]
I adjusted the change here, I think @akien-mga was right here. It seems that MeshInstance3D::blend_shape_properties is mainly a wrapper for the track values, and can be cleared upon a mesh change regardless. I tested around and it seems to save the values of the blendshapes when changing from suzanne1 to 2, and removes the extra ones when changing from 3 to 1. 

I think a part of the issue was that the properties hashmap was never getting cleared and then when we generate the properties list, it's referencing values which don't exist anymore in the mesh.

**Review by clayjohn — DISMISSED** [review]
This makes sense!

**Comment by Lielay9** [comment]
Umm, that does look like a correct place for clearing the `blend_shape_properties` but shouldn't the `blend_shape_tracks` still need to be cleared in the set_mesh function when swapping the mesh to an another?



The value for blendshape index 0 remains even after changing the mesh to an another. (Note the meshes are different despite looking the same).

Same in code:
```gdscript
mesh = preload("res://suzanne2.res")
set_blend_shape_value(0, 1.0)
printt(get_blend_shape_value(0), "Blendshape value set to 1.0") # 1.0
mesh = preload("res://suzanne3.res")
printt(get_blend_shape_value(0), "Blendshape should probably be 0 after changing mesh.") # 1.0
```

**Comment by deppong** [comment]
@Lielay9 You are 100% right! I updated the patch a bit, and I'll get to fixing any styling checks that fail, since I'm sure I made some errors being new to the codebase.

The fix here basically saves all of the previous property names into a Vector<String> before then clearing the blend_shape_properties HashMap. After that instead of using the initialize_bs_from integer to set the values, we check if the current new meshes blend shape name is in our save Vector. If so then we can go ahead and use the old value. if not we set it to zero, and then assign our property name.

After doing more testing it seems like the values do get properly cleared out as well, and no errors happen. The old values are preserved if the name is the same, and any old values and properties not present in the new mesh are cleared.

**Comment by Lielay9** [comment]
> @Lielay9 You are 100% right! I updated the patch a bit, and I'll get to fixing any styling checks that fail, since I'm sure I made some errors being new to the codebase.
> 
> The fix here basically saves all of the previous property names into a Vector before then clearing the blend_shape_properties HashMap. After that instead of using the initialize_bs_from integer to set the values, we check if the current new meshes blend shape name is in our save Vector. If so then we can go ahead and use the old value. if not we set it to zero, and then assign our property name.
> 
> After doing more testing it seems like the values do get properly cleared out as well, and no errors happen. The old values are preserved if the name is the same, and any old values and properties not present in the new mesh are cleared.

I'm sorry I wasn't clear enough. The blendshape values should **not** be retained when the mesh is swapped even if the blendshape names happen to match. i.e. the `blend_shape_tracks` should be cleared in the `set_mesh()` function before calling `_mesh_changed()`. (The way you did in the first iteration).

Furthermore, I think what you had in the second iteration (+the above), was enough, no extra needed.

Explanation: If the `_mesh_changed()` is called after setting the mesh it means the mesh resource changed (no swapping, the resource itself changed, vertex moved etc., English is confusing) and if the new and old blendshapes don't match, you're screwed anyway: there's no way to know if the order was changed (which you fixed and didn't work before) or if one was renamed(Worked before but wouldn't work after this pr). So I'd suggest against this change since it isn't necessarily a fix and there's a trade-off (even though it's arguably favorable). AND do note that the importing already worked without this change, so it would really only affect folks editing meshes in Godot.

In short, what I'd recommend is to: Combine your first and second iterations i.e. clear  `blend_shape_tracks` in `set_mesh()` and `blend_shape_properties` in `mesh_changed()`, and leave the renaming/value-retaining logic otherwise as is.

**Comment by deppong** [comment]
Ah, I understand now sorry @Lielay9! I can make the changes, I was a bit confused, since I was also following what @akien-mga advised here:

> I read the previous code as trying to preserve the configuration of blend shapes when updating a mesh with a new version of the same mesh (e.g. if you re-exported the mesh from Blender). The change in this PR might mean that such use case would no longer be supported, which is arguably a worse issue than the bug it's fixing.

Let me know if I should add it as a separate commit so we can pick between the two behaviors after some more people chime in on the preferred way of going about it!

**Comment by Lielay9** [comment]
> Ah, I understand now sorry @Lielay9! I can make the changes, I was a bit confused, since I was also following what @akien-mga advised here:
> 
> > I read the previous code as trying to preserve the configuration of blend shapes when updating a mesh with a new version of the same mesh (e.g. if you re-exported the mesh from Blender). The change in this PR might mean that such use case would no longer be supported, which is arguably a worse issue than the bug it's fixing.

Yes they had a concern that that **could** happen but, sorry again for not having been clear enough, I specifically tested that what they speculated, is not a problem. From before:

> After Patch:
>
> * Import Inherited scene -> Values preserved by name (Good)

I've may have phrased that poorly but that's exactly the result of _"updating a mesh with a new version of the same mesh (e.g. if you re-exported the mesh from Blender)."_ Which once again did work even without the latest changes; changing the 2 lines was enough and did not impact how thing are re-imported.

**Comment by deppong** [comment]
@Lielay9 Thanks again so much for the clarification, I think may have also misinterpreted what you had said. I `commit --amend`ed the patch now with all of that in order. Sorry about the mixup! 

Let me know if there is anything else I'm missing for this patch.

**Comment by Lielay9** [comment]
Looks good, only a small nitpick: the latter line in else can be removed. (Github won't let me suggest change; [github pls fix](https://github.com/github/roadmap/issues/347))
https://github.com/godotengine/godot/blob/a3b0e0f4cc3ed188440f2f51a3b408660e75138e/scene/3d/mesh_instance_3d.cpp#L126-L127
I'll do sanity check and run the test project again after getting some sleep, but otherwise should be good.

**Comment by Lielay9** [comment]
TLDR. Everything works™

Orignal mesh (suzanne.res):


Reimporting the same mesh with different blendshapes preserves the values if the name matches (Tested by editing the mesh in blender and reimporting):

After removing blendshape:


After changing blendshape name:


Swapping mesh in editor clears previous blendshapes even if the name matches:

Before swap:


After swapping to a mesh with blendshape that matches:


And finally swapping mesh in code also clears blendshapes properly:
```gdscript
extends MeshInstance3D

func _ready() -> void:
	mesh = preload("res://suzanne.res")
	set_blend_shape_value(0, 1.0)
	set_blend_shape_value(1, 1.0)
	print(mesh.get_blend_shape_name(0)) # Frown
	print(mesh.get_blend_shape_name(1)) # Smile
	print(get_blend_shape_value(1)) # 1.0
	mesh = preload("res://notsuzanne.res")
	print(mesh.get_blend_shape_name(0)) # Smile
	print(get_blend_shape_value(0)) # 0.0
```

**Review by clayjohn — CHANGES_REQUESTED** [review]
I don't think this is correct anymore. I have a feeling this is only working by relying on behaviour of memdelete that is not guaranteed. 

With this PR when setting a new mesh, we do the following:
1. Clear `blend_shape_tracks` this frees the underlying memory (but does not zero it)
2. Call `mesh_changed`
3. Resize `blend_shape_tracks` this allocates memory (but does not zero it)
4. Reset the blend shape value to the value in `blend_shape_tracks`

But we freed that memory! This PR relies on the `blend_shape_tracks` re-allocating from the same chunk of memory that it had just deallocated from plus not clearing the memory during deletion or re-allocation. This is a classic use-after-free error.

**Comment by Lielay9** [comment]
> I don't think this is correct anymore. I have a feeling this is only working by relying on behaviour of memdelete that is not guaranteed.
> 
> With this PR when setting a new mesh, we do the following:
> 
>     1. Clear `blend_shape_tracks` this frees the underlying memory (but does not zero it)
> 
>     2. Call `mesh_changed`
> 
>     3. Resize `blend_shape_tracks` this allocates memory (but does not zero it)
> 
>     4. Reset the blend shape value to the value in `blend_shape_tracks`
> 
> 
> But we freed that memory! This PR relies on the `blend_shape_tracks` re-allocating from the same chunk of memory that it had just deallocated from plus not clearing the memory during deletion or re-allocation. This is a classic use-after-free error.

It's fine, no? Isn't `initialize_bs_from` specifically for that.
1. `blend_shape_tracks` is cleared (memory freed, size is '0')
2. call `mesh_changed`
3. `initialize_bs_from` is set to `blend_shape_tracks.size()` (which is '0')
4. `blend_shape_tracks` is resized and allocates memory (size = blendshape-count in new mesh)
6. if i < initialize_bs_from set the value from blend_shape_tracks (which won't happen since initialize_bs_from is '0') else initialize the value to '0'

So even if `blend_shape_tracks` contains uninitilized values there momentarily, only the ones that are previously set will be used, and new ones (i >= `initialize_bs_from`) are initilized to 0.

Or do you mean that when `blend_shape_tracks` is resized the values it had before (provided they fit) won't carry over?

**Comment by deppong** [comment]
> I don't think this is correct anymore. I have a feeling this is only working by relying on behaviour of memdelete that is not guaranteed.
> 
> With this PR when setting a new mesh, we do the following:
> 
>     1. Clear `blend_shape_tracks` this frees the underlying memory (but does not zero it)
> 
>     2. Call `mesh_changed`
> 
>     3. Resize `blend_shape_tracks` this allocates memory (but does not zero it)
> 
>     4. Reset the blend shape value to the value in `blend_shape_tracks`
> 
> 
> But we freed that memory! This PR relies on the `blend_shape_tracks` re-allocating from the same chunk of memory that it had just deallocated from plus not clearing the memory during deletion or re-allocation. This is a classic use-after-free error.

 So I am a bit confused here, it doesn't look like _mesh_changed() is really called anywhere outside of MeshInstance3D::set_mesh, which wouldn't that make the need for step 4. here not needed? In some testing, it seems like things are behaving the same way when the uses of `initialize_bs_from` are removed, even reimporting.

**Comment by Lielay9** [comment]
> So I am a bit confused here, it doesn't look like _mesh_changed() is really called anywhere outside of MeshInstance3D::set_mesh, which wouldn't that make the need for step 4 not necessary? We can just zero initialize the tracks in the newly reallocated memory, right? In some testing, it seems like things are behaving the same way when the uses of `initialize_bs_from` are removed, even reimporting.
> 
> Please let me know if I'm mistaken on this, or if my thinking here is flawed.

It is connected to the changed callback in set_mesh:
https://github.com/godotengine/godot/blob/6681f2563b99e14929a8acb27f4908fece398ef1/scene/3d/mesh_instance_3d.cpp#L123

So it is called whenever the mesh-resource is updated: 
```gdscript
var am: ArrayMesh = preload("res://suzanne.res")
mesh = am # set_mesh() is called and '_mesh_changed()' is connected to 'changed'
am.emit_changed() # Manually triggers the callback; not very useful of course.
# A more realistic scenario:
set_blend_shape_value(0, 1.0) # Set 'Smile' value to 1.0
am.set_blend_shape_name(0, "Not smile") # Change name: Smile -> Not Smile; Triggers the callback.
print(get_blend_shape_value(0)) # Value of 'Not Smile' is 1.0 (since values are preserved by index).
```

**Comment by deppong** [comment]
> It is connected to the 'changed' in set_mesh:
> 
> https://github.com/godotengine/godot/blob/6681f2563b99e14929a8acb27f4908fece398ef1/scene/3d/mesh_instance_3d.cpp#L123

Thanks for the clarification! I kinda figured that's what the above callback system did, I just wasn't sure. It does seem that when you are directly calling it, the original memory (tracks) isn't freed, so there isn't an issue with a UAF, and that variable helps control that from the set_mesh() function we have freed it to prevent the memory from being accessed. 

@clayjohn let us know if our thinking is correct here...
